File: dotnet/1-50/15.cs
using System;

public class Program{
	const int mod = 1000000007;
	public static int BinPow(int a, int n){
      	if (n == 0) return 1;
      	if (n % 2 == 0){
          	int x = BinPow(a, n / 2);
          	return (int)(((long)x * x) % mod);
      	}
      	return (int)(((long)a * BinPow(a, n - 1)) % mod);
  	}
	
	public static void Main(){
      int n, k, a, b;
      string[] input = Console.ReadLine().Split();
      n = int.Parse(input[0]);
      k = int.Parse(input[1]);

      a = BinPow(k, n) - 1;
      b = k - 1;
      int result = (int)(((long)a * BinPow(b, mod - 2)) % mod);
      Console.WriteLine(result);
	}
}

File: Swift/1-50/15.swift
let mod = Int(1e9 + 7)

func binpow(_ a: Int, _ n: Int) -> Int {
    if n == 0 {
        return 1
    }
    if n % 2 == 0 {
        let x = binpow(a, n / 2)
        return (1 * x * x) % mod
    }
    return (1 * a * binpow(a, n - 1)) % mod
}

if let input = readLine() {
    let nk = input.split(separator: " ").map { Int($0)! }
    let n = nk[0]
    let k = nk[1]

    let a = binpow(k, n) - 1
    let b = k - 1

    let result = (1 * a * binpow(b, mod - 2)) % mod
    print(result)
}
